stan/math/opencl/matrix_cl.hpp

```diff
@@ -220,6 +211,125 @@ class matrix_cl {
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, vari_size);
+        cl::Buffer oclBuffer_adj
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, vari_size);
+        // We want to do a single loop over the vari so we use the map
+        // method to copy the data over.
+        char* val_ptr = (char*)queue.enqueueMapBuffer(
+            oclBuffer_val, CL_MAP_READ | CL_MAP_WRITE, CL_MEM_READ_WRITE, 0,
+            vari_size);
+        char* adj_ptr = (char*)queue.enqueueMapBuffer(
+            oclBuffer_adj, CL_MAP_READ | CL_MAP_WRITE, CL_MEM_READ_WRITE, 0,
+            vari_size);
+
+        for (std::size_t i = 0; i < ((R * C) / 2); i++) {
+          val_ptr[i] = A[i]->val_;
+          adj_ptr[i] = A[i]->adj_;
+        }
+        queue.enqueueUnmapMemObject(oclBuffer_val, val_ptr);
+        queue.enqueueUnmapMemObject(oclBuffer_adj, adj_ptr);
+        val_ = matrix_cl(oclBuffer_val, R, C);
+        adj_ = matrix_cl(oclBuffer_adj, R, C);
+      } catch (const cl::Error& e) {
+        check_opencl_error("matrix constructor", e);
+      }
+    }
+  }
+  /**
+   * Constructor for the matrix_cl that
+   * creates a copy of a var type Eigen matrix on the GPU.
+   *
+   *
+   * @tparam R rows of matrix
+   * @tparam C cols of matrix
+   * @param A the Eigen matrix
+   *
+   * @throw <code>std::system_error</code> if the
+   * matrices do not have matching dimensions
+   */
+  template <int R, int C>
+  explicit matrix_v_cl(const Eigen::Matrix<var, R, C>& A)
+      : rows_(A.rows()), cols_(A.cols()) {
+    cl::Context& ctx = opencl_context.context();
+    cl::CommandQueue& queue = opencl_context.queue();
+    if (size() > 0) {
+      try {
+        const int vari_size = sizeof(double) * ((R * C) / 2);
+        // creates the OpenCL buffer to copy the Eigen
+        // matrix to the OpenCL device
+        cl::Buffer oclBuffer_val
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, vari_size);
+        cl::Buffer oclBuffer_adj
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, vari_size);
+
+        // We want to do a single loop over the vari so we use the map
+        // method to copy the data over.
+        char* val_ptr
+            = queue.enqueueMapBuffer(oclBuffer_val, CL_MAP_READ | CL_MAP_WRITE,
+                                     CL_MEM_READ_WRITE, 0, vari_size);
+        char* adj_ptr
+            = queue.enqueueMapBuffer(oclBuffer_adj, CL_MAP_READ | CL_MAP_WRITE,
+                                     CL_MEM_READ_WRITE, 0, vari_size);
+
+        for (std::size_t i = 0; i < A.size(); i++) {
+          val_ptr[i] = A.coeffRef(i).vi_->val_;
+          adj_ptr[i] = A.coeffRef(i).vi_->adj_;
+        }
+        queue.enqueueUnmapMemObject(oclBuffer_val, val_ptr);
+        queue.enqueueUnmapMemObject(oclBuffer_adj, adj_ptr);
+        val_ = matrix_cl(oclBuffer_val, R, C);
+        adj_ = matrix_cl(oclBuffer_adj, R, C);
+      } catch (const cl::Error& e) {
+        check_opencl_error("matrix constructor", e);
+      }
+    }
+  }
+};
```

Comment:
@seantalts @rok-cesnovar curious what you guys think of the above? It's not tested yet, but is mostly readable and compiles. I made a new class `matrix_v_cl` for a few reasons

1. I thought having the cache for some constructors and not others in the `matrix_cl` code was kind of odd. I think the above is more clear. Handing a `matrix_cl` or buffer will not cause the cache to go off. Handing it an Eigen matrix of doubles will cause it to go off.
2. Composition over inheritance. The other styles I tried were either kind of gross inheritance or some other wonkish stuff
3. I think the code will end up looking a little nicer. So for instance in the cholesky code we can do something like
```c++
matrix_v_cl L(vari);
// doing math now like
L.adj_ * L.val_;
```
4. With this we use the `map` style of moving memory over. This is usually the same as the write and read functions, but in the case of grabbing both parts of the `vari` we get contiguous memory

Reply:
I like it a lot on first glance. I also love the use of `enqueueMap`, it would be great to see if we could use it everywhere, potentially together with host pointers etc. But that is another PR. 

I am going to try and fix the lint errors and doxygen stuff to see if this passes all the tests.

Reply:
Rok have you ever seen performance improvements using enqueueMap? Everyone on the internet says it's faster, but when I've tried it with a few different flags I've never seen a big speed diff

Reply:
Seems pretty good to me! I like the design principles you used. Do we always know when it's going to be a var matrix in e.g. heavily templated functions like the probability distributions? That might be an argument for a templated construction function that specializes to either matrix_v_cl or matrix_cl under the hood. In either case it would seem that the Eigen constructor could outsource most of the effort to the vari** ctor?

A var matrix never gets cached, right? Is this new class related to the caching mechanism somehow?

Reply:
@SteveBronder if you go with enqueueMap you have to use the host pointers and non blocking calls in order to see a difference. As far as I know and understand it loads the buffer when it needs it hence the speedup (you get the stuff from out of order execution for free I guess). Havent tried it myself, but would really like to. It sounds like the dream :dancer: 

The biggest boost of enqueueMap is actually if you use CPUs, because no buffer copy occurs then,  (or at least that is what they promise) and the overhead is minimum.

Reply:
> Seems pretty good to me! I like the design principles you used. Do we always know when it's going to be a var matrix in e.g. heavily templated functions like the probability distributions? 

Hmm, idk? Like I'm looking at [`multi_normal_cholesky_lpdf`](https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/prob/multi_normal_cholesky_lpdf.hpp). Possibly? Lot going on there but could be not good.

> That might be an argument for a templated construction function that specializes to either matrix_v_cl or matrix_cl under the hood. 

I think I get what your laying down here. In that case I would probs have a `matrix_cl` that then calls something matrix_v_cl and matrix_d_cl under the hood. 

> In either case it would seem that the Eigen constructor could outsource most of the effort to the vari** ctor?

Not sure if I'm getting this piece

> A var matrix never gets cached, right? Is this new class related to the caching mechanism somehow?

Yes, so the matrix_v_cl is because we don't want to cache the var and I'm trying to work out the nicest way to represent that

Reply:
I can add the `enqueueMap` stuff here since I think I get the pattern. For now we can set it as blocking, then once I finish up the async stuff we can get some big ol' speedups :)

Reply:
@rok-cesnovar I got the map/unmap thing to work, but only with `CL_MEM_USE_HOST_PTR` while `CL_MEM_ALLOC_HOST_PTR` is the one that gives pinned memory. So I'm going to shelve that whole idea until we start working on the async stuff

https://gist.github.com/SteveBronder/2ed15f5327490a1cabdcd607f50fc7e2#file-stan_var_matrix_cl-hpp

Reply:
Ah too bad. Lets shelve that yeah. Thanks for the info!